### PR TITLE
implement hashCode for complex datatype.

### PR DESCRIPTION
**Pull request by svmundada**
1. Complex number hashing is taken from cpython's [hashing function](https://github.com/python/cpython/blob/master/Objects/complexobject.c#L408).

2. I made __repr__ and __str__ methods same for complex numbers, I am not exactly sure if it is the right thing but from here cypthon ```complexobject.c``` module both seem to have same underlining function.
Link to cpython's corressponding functions:
[```__repr__```](https://github.com/python/cpython/blob/master/Objects/complexobject.c#L1126) and [```__str__```](https://github.com/python/cpython/blob/master/Objects/complexobject.c#L1132).

Thank you.

### CODE REVIEW

File: python/common/org/python/types/Complex.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
     }
 
     public int hashCode() {
-        return this.hashCode();
+        return this.real.hashCode() + 0xf4243 * this.imag.hashCode();
```
Comment (freakboy3742) [inline python/common/org/python/types/Complex.java]: What's the significance of 0xf4342? It feels like a magic number - was it a randomly chosen magic number, or does it have history in the CPython codebase? Whatever the explanation, this warrants a comment in the code to explain where the number came from, and what it's being used for.

File: python/common/org/python/types/Complex.java [hunk 2]
```diff
@@ -162,31 +162,11 @@ private String partToStr(org.python.types.Float x) {
...
-        } else {
-            real_present = false;
-        }
-        if (this.real.value != 0 && this.imag.value >= 0) {
-            buffer.append("+");
-        }
-        if (((org.python.types.Bool) ((this.imag).__int__().__eq__(this.imag))).value) {
-            buffer.append(((org.python.types.Str) (this.imag).__int__().__repr__()).value);
+        if (this.real.value != 0.0 || this.real.isNegativeZero()) {
+            return new org.python.types.Str("(" + partToStr(this.real) + ((this.imag.value >= 0.0 && !this.imag.isNegativeZero()) ? "+" : "-") + partToStr(new org.python.types.Float(Math.abs(this.imag.value))) + "j)");
```
Comment (freakboy3742) [inline python/common/org/python/types/Complex.java]: I'm surprised this wasn't picked up by linting... but we don't use ternary operators, and the line length is *extremely* excessive.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Can't argue with the results, but there's some comments and style issues that need to be addressed.

File: python/common/org/python/types/Complex.java (hunk 2 above)
Comment (student-06) [reply python/common/org/python/types/Complex.java]: Actually we are using ternary at many places in voc, so should we remove all of those ? I'll make that last line bit more readable.

Comment (freakboy3742) [reply python/common/org/python/types/Complex.java]: We shouldn't be - and the linter *should* be catching them. If it isn't, then we have another problem.

**Comment by student-06** [comment]
@freakboy3742 could you please check the [new](https://github.com/pybee/voc/pull/755/commits/20c92c92b91016de0204fdf6a20263a71d850636) commit.
Thank you.

**Comment by phildini** [comment]
Hi there! It looks like this PR might be dead, so we're closing it for now. Feel free to re-open it if you'd like to continue, or think about directing your efforts to https://github.com/beeware/briefcase or https://github.com/beeware/toga. Both of these have more active development right now. 😄

**Pull request closed without merging**